torch/distributions/constraints.py

```diff
@@ -1,5 +1,20 @@
-import torch
+r"""
+The following constraints are implemented:
+
+- ``constraints.dependent``
+- ``constraints.boolean``
+- ``constraints.greater_than(lower_bound)``
+- ``constraints.integer_interval(lower_bound, upper_bound)``
+- ``constraints.interval(lower_bound, upper_bound)``
+- ``constraints.lower_triangular``
+- ``constraints.nonnegative_integer``
+- ``constraints.positive``
+- ``constraints.real``
+- ``constraints.simplex``
+- ``constraints.unit_interval``
+"""
 
+import torch
 
 __all__ = [
     'Constraint',
```

Comment:
This is a bike-shedding question, but `Constraints` also incorporates unconstrained spaces like `real`. Is there a name conflict with other parts of PyTorch's API if we rename `Constraint` to `Type`? (maybe tensor type?)

If I understand correctly, the idea is to formalize the category that a value or object belongs to. `Type` captures this, and also connects to Python type annotations if we ever plan to annotate manipulations like a Bijector or Transform's domain/range.

Reply:
After conversation with @pravnar at PPS 2018, I guess another possible term is "base measure". https://pps2018.soic.indiana.edu/files/2017/12/abstract.pdf

(I'm personally happy to defer to whatever, `Constraints` included—only bringing up other possibilities.)

Reply:
Hmm base measure makes sense for `.support` but I think it assumes too much structure for non-Bayesian treatment of parameters. E.g. in maximum likelihood there is no measure.

I kindof prefer `Constraint` rather than `Type` out of fear that these objects might attract additional type-like information (bit precision, cuda device, dimension, reparameterization type, is_exponential_family, ...). `Constraint` really emphasizes the non-goals :smile:

Reply:
Oh @pravnar's approach might be useful for automatically inferring infinite KL's from base measure, as @vishwakftw and I have been discussing in #77

Reply:
+1 to keeping `Constraint` and using it for infinite KLs!